### PR TITLE
Support mathematical programming syntax parsing

### PR DESCRIPTION
**Pull request by typhoonzero**
Part of https://github.com/sql-machine-learning/sqlflow/issues/2403

### CODE REVIEW

File: pkg/parser/extended_syntax_parser.y [hunk 1]
```diff
@@ -62,12 +62,14 @@ type SQLFlowSelectStmt struct {
 	Predict  bool
 	Explain  bool
 	Evaluate bool
+	Mathp    bool
```
Comment (wangkuiyi) [inline pkg/parser/extended_syntax_parser.y]: Mathp => Optimize

File: pkg/parser/extended_syntax_parser.y [hunk 2]
```diff
@@ -62,12 +62,14 @@ type SQLFlowSelectStmt struct {
...
 	ShowTrain bool
 	StandardSelect
 	TrainClause
 	PredictClause
 	ExplainClause
 	EvaluateClause
+	MathProgClause
```
Comment (wangkuiyi) [inline pkg/parser/extended_syntax_parser.y]: MathProgClause => OptimizeClause

File: pkg/parser/extended_syntax_parser.y [hunk 3]
```diff
@@ -110,6 +112,16 @@ type EvaluateClause struct {
 	EvaluateInto  string
 }
 
+type MathProgClause struct {
+	// Sense can be MAXIMIZE or MINIMIZE
+	Sense string
```
Comment (wangkuiyi) [inline pkg/parser/extended_syntax_parser.y]: Sense string => Direction

File: pkg/parser/extended_syntax_parser.y [hunk 4]
```diff
@@ -110,6 +112,16 @@ type EvaluateClause struct {
...
+	Objective *Expr
+	Constrants ExprList
+	SolveAttrs Attributes
+	Solver string
+	SolveResult string
```
Comment (wangkuiyi) [inline pkg/parser/extended_syntax_parser.y]: SolveResult => Result

File: pkg/parser/extended_syntax_parser.y [hunk 5]
```diff
@@ -110,6 +112,16 @@ type EvaluateClause struct {
 	EvaluateInto  string
 }
 
+type MathProgClause struct {
+	// Sense can be MAXIMIZE or MINIMIZE
+	Sense string
+	Objective *Expr
+	Constrants ExprList
+	SolveAttrs Attributes
```
Comment (wangkuiyi) [inline pkg/parser/extended_syntax_parser.y]: I don't get what are Solve Attributes. Is it SolverAttrs?

File: pkg/parser/extended_syntax_parser.y [hunk 6]
```diff
@@ -142,6 +154,7 @@ func attrsUnion(as1, as2 Attributes) Attributes {
   infr PredictClause
   expln ExplainClause
   evalt EvaluateClause
+  mathp MathProgClause
```
Comment (wangkuiyi) [inline pkg/parser/extended_syntax_parser.y]: mathp MathProgClause => optim OptimizeClause

File: pkg/parser/extended_syntax_parser.y [hunk 7]
```diff
@@ -153,13 +166,14 @@ func attrsUnion(as1, as2 Attributes) Attributes {
 %type  <infr> predict_clause
 %type  <expln> explain_clause
 %type  <evalt> evaluate_clause
+%type <mathp> mathprog_clause
```
Comment (wangkuiyi) [inline pkg/parser/extended_syntax_parser.y]: mathprog_clause => optimize_clause

File: pkg/parser/extended_syntax_parser.y (hunk 5 above)
Comment (typhoonzero) [reply pkg/parser/extended_syntax_parser.y]: Changed to `OptimizeAttrs`

File: pkg/parser/extended_syntax_parser.y (hunk 4 above)
Comment (typhoonzero) [reply pkg/parser/extended_syntax_parser.y]: Changed to `OptimizeInto` to be the same format of other clause structs like `EvaluateClause`.

**Review by sneaxiy — DISMISSED** [review]
LGTM++

File: pkg/parser/extended_syntax_parser.y [hunk 8]
```diff
@@ -62,12 +62,14 @@ type SQLFlowSelectStmt struct {
 	Predict  bool
 	Explain  bool
 	Evaluate bool
+	Optimize    bool
```
Comment (shendiaomo) [inline pkg/parser/extended_syntax_parser.y]: Redundant whitespaces

**Review by shendiaomo — CHANGES_REQUESTED** [review]

**Review by shendiaomo — APPROVED** [review]

**Pull request merged**